mars/tensor/einsum/core.py

```diff
@@ -0,0 +1,264 @@
+            axes_splits.update(axis_splits)
+            tensor_contract_axes.append([ax for ax in axes if ax not in output_scripts])
+
+        out_chunks = []
+        output_indexes = [tensor_indexes[ax] for ax in output_scripts]
+        for out_idx in itertools.product(*output_indexes):
+            all_indexes = [[None] * t.ndim for t in input_tensors]
+            tensor_shape = []
+            for i, idx in enumerate(out_idx):
+                tensor_shape.append(axes_splits[output_scripts[i]][idx])
+                for (t_idx, axis) in output_axes[output_scripts[i]]:
+                    if input_tensors[t_idx].shape[axis] == 1:
+                        all_indexes[t_idx][axis] = 0
+                    else:
+                        all_indexes[t_idx][axis] = idx
+            tensor_shape = tuple(tensor_shape)
+            einsum_chunks = []
+            contract_axes = [s for s in set(input_scripts.replace(',', '')) if s not in output_scripts]
+            for contract_indexes in itertools.product(*[range(len(axes_splits[ax])) for ax in contract_axes]):
+                for j, t_contract_axes in enumerate(tensor_contract_axes):
+                    for axis in t_contract_axes:
+                        axis_index = tensor_axes[j][1].index(axis)
+                        all_indexes[j][axis_index] = contract_indexes[contract_axes.index(axis)]
+                einsum_op = op.copy().reset_key()
+                in_chunks = [t.cix[tuple(indices)] for t, indices in zip(input_tensors, all_indexes)]
+                chunk = einsum_op.new_chunk(in_chunks, shape=tensor_shape, order=out_tensor.order)
+                einsum_chunks.append(chunk)
+
+            if len(einsum_chunks) == 1:
+                c = einsum_chunks[0]
+                chunk_op = c.op.copy()
+                chunk = chunk_op.new_chunk(c.inputs, shape=c.shape, index=out_idx, order=out_tensor.order)
+            else:
+                chunk = tree_add(op.dtype, einsum_chunks, out_idx, tensor_shape, sparse=op.sparse)
+            out_chunks.append(chunk)
+
+        nsplits = [axes_splits[ax] for ax in output_scripts]
+        new_op = op.copy()
+        return new_op.new_tensors(input_tensors, out_tensor.shape, chunks=out_chunks, nsplits=nsplits)
+
+    @classmethod
+    def execute(cls, ctx, op):
+        inputs, device_id, xp = as_same_device(
+            [ctx[c.key] for c in op.inputs], device=op.device, ret_extra=True)
+
+        with device(device_id):
+            if xp is np:
+                ctx[op.outputs[0].key] = xp.einsum(op.subscripts, *inputs, optimize=op.optimize, dtype=op.dtype,
+                                                   order=op.order, casting=op.casting)
+            else:
+                # Cupy doesn't support `optimize`, `order` and `casting`.
+                ctx[op.outputs[0].key] = xp.einsum(op.subscripts, *inputs, dtype=op.dtype)
+
+
+def einsum(subscripts, *operands, dtype=None, order='K', casting='safe', optimize=False):
+    """
```

Comment:
Just copy the doc from numpy, examples needed as well.

Reply:
Refer to https://github.com/numpy/numpy/blob/d9b1e32cb8ef90d6b4a47853241db2a28146a57d/numpy/core/einsumfunc.py#L1005

Reply:
OK, I will add the part of examples.